File: sdks/csharp/src/NovaBilling/Types/CouponResponse.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record CouponResponse : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    [JsonPropertyName("id")]
    public required string Id { get; set; }

    [JsonPropertyName("code")]
    public required string Code { get; set; }

    [JsonPropertyName("name")]
    public required string Name { get; set; }

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("discountType")]
    public required CouponResponseDiscountType DiscountType { get; set; }

    /// <summary>
    /// Discount value as decimal string
    /// </summary>
    [JsonPropertyName("discountValue")]
    public required string DiscountValue { get; set; }

    [JsonPropertyName("currency")]
    public string? Currency { get; set; }

    [JsonPropertyName("maxRedemptions")]
    public double? MaxRedemptions { get; set; }

    [JsonPropertyName("redemptionCount")]
    public required double RedemptionCount { get; set; }

    [JsonPropertyName("appliesToPlanIds")]
    public IEnumerable<string> AppliesToPlanIds { get; set; } = new List<string>();

    [JsonPropertyName("isActive")]
    public required bool IsActive { get; set; }

    [JsonPropertyName("expiresAt")]
    public string? ExpiresAt { get; set; }

    [JsonPropertyName("createdAt")]
    public required string CreatedAt { get; set; }

    [JsonPropertyName("updatedAt")]
    public required string UpdatedAt { get; set; }

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}

File: sdks/ruby/lib/novabilling/types/coupon_response.rb
# frozen_string_literal: true

module Novabilling
  module Types
    class CouponResponse < Internal::Types::Model
      field :id, -> { String }, optional: false, nullable: false
      field :code, -> { String }, optional: false, nullable: false
      field :name, -> { String }, optional: false, nullable: false
      field :description, -> { String }, optional: true, nullable: false
      field :discount_type, -> { Novabilling::Types::CouponResponseDiscountType }, optional: false, nullable: false, api_name: "discountType"
      field :discount_value, -> { String }, optional: false, nullable: false, api_name: "discountValue"
      field :currency, -> { String }, optional: true, nullable: false
      field :max_redemptions, -> { Integer }, optional: true, nullable: false, api_name: "maxRedemptions"
      field :redemption_count, -> { Integer }, optional: false, nullable: false, api_name: "redemptionCount"
      field :applies_to_plan_ids, -> { Internal::Types::Array[String] }, optional: false, nullable: false, api_name: "appliesToPlanIds"
      field :is_active, -> { Internal::Types::Boolean }, optional: false, nullable: false, api_name: "isActive"
      field :expires_at, -> { String }, optional: true, nullable: false, api_name: "expiresAt"
      field :created_at, -> { String }, optional: false, nullable: false, api_name: "createdAt"
      field :updated_at, -> { String }, optional: false, nullable: false, api_name: "updatedAt"
    end
  end
end
